python/revng/cli/_commands/idb_converter.py

```diff
@@ -0,0 +1,655 @@
+
+            # Emit final aggregated member.
+            if accumulated_bitfield_width > 0:
+                aggregated_member = emit_aggregated_member(
+                    aggregate_field_underlying_size,
+                    committed_size,
+                    aggregate_fields_count,
+                )
+                fields.append(aggregated_member)
+                committed_size += aggregate_field_underlying_size
+
+            revng_type.Fields = fields
+            revng_type.Size = committed_size
+
+    def _fixup_unions(self):
+        while self._unions_to_fixup:
+            revng_type, idb_type = self._unions_to_fixup.pop()
+
+            assert isinstance(revng_type, m.UnionType)
+            assert idb_type.is_decl_union()
+
+            fields = []
+            for idx, member in enumerate(idb_type.type_details.members):
+                qualified_type = self._convert_idb_type_to_revng_type(member.type)
+                revng_member = m.UnionField(
+                    OriginalName=member.name,
+                    Type=qualified_type,
+                    Index=idx,
+                )
+                fields.append(revng_member)
+
+            revng_type.Fields = fields
+
+    def _convert_idb_type_to_revng_type(
+        self,
+        type: idb.typeinf.TInfo,  # noqa: A002
+        ordinal=None,
+    ) -> m.QualifiedType:
+        """Converts the given TInfo obtained from python-idb to the corresponding revng
+        QualifiedType. If available, the integer identifying the type in the IDB (ordinal) should
+        be supplied to allow handling circular references. If a type with the given ordinal was
+        already converted the same instance is returned.
+        """
+        assert isinstance(type, idb.typeinf.TInfo)
+
+        # Check if we already converted this type, and if so return the existing type.
+        # Fundamental to handle circular dependencies.
+        existing_revng_type = self.idb_types_to_revng_types.get(ordinal)
+        if existing_revng_type is not None:
+            return existing_revng_type
+
+        type_name = type.get_name()
+        revng_type_qualifiers = []
+
+        if type.is_decl_typedef():
+            aliased_type = type.get_next_tinfo()
+
+            # Special case reserved typedefs
+            match = re.fullmatch(r"(?P<unsigned>u)?int(?P<width>(8|16|32|64))_t", type_name)
+            if match:
+                unsigned = bool(match["unsigned"])
+                size = int(match["width"]) // 8
+                primitive_kind = (
+                    m.PrimitiveTypeKind.Unsigned if unsigned else m.PrimitiveTypeKind.Signed
+                )
+                qualified_type = self._get_primitive_type(primitive_kind, size)
+            elif type_name == "size_t":
+                primitive_kind = m.PrimitiveTypeKind.Unsigned
+                size = 8 if self.is64bit else 4
+                qualified_type = self._get_primitive_type(primitive_kind, size)
```

Comment:
Ouch... it would generate:

```
Types:
- ID: 1025
  Kind: PrimitiveType
  PrimitiveKind: Number
  Size: 1
- ID: 1284
  Kind: PrimitiveType
  PrimitiveKind: Unsigned
  Size: 4
- ID: 6940297126226104124
  Kind: TypedefType
  OriginalName: uint32_t
  UnderlyingType:
    Qualifiers: []
    UnqualifiedType: /Types/PrimitiveType-1284
- ID: 9514706589701521359
  Kind: TypedefType
  OriginalName: uint32_t
  UnderlyingType:
    Qualifiers: []
    UnqualifiedType: /Types/PrimitiveType-1284
- ABI: SystemV_x86_64
  Arguments:
  - Index: 0
    OriginalName: x
    Type:
      Qualifiers: []
      UnqualifiedType: /Types/TypedefType-9514706589701521359
  ID: 10813545522046549192
  Kind: CABIFunctionType
  OriginalName: fn
  ReturnType:
    Qualifiers: []
    UnqualifiedType: /Types/TypedefType-6940297126226104124
```

instead of:

```
Types:
- ID: 1025
  Kind: PrimitiveType
  PrimitiveKind: Number
  Size: 1
- ID: 2692631726386566055
  Kind: TypedefType
  OriginalName: uint32_t
  UnderlyingType:
    Qualifiers: []
    UnqualifiedType: /Types/PrimitiveType-1025
- ID: 10726857336666412012
  Kind: TypedefType
  OriginalName: uint32_t
  UnderlyingType:
    Qualifiers: []
    UnqualifiedType: /Types/PrimitiveType-1025
- ABI: SystemV_x86_64
  Arguments:
  - Index: 0
    OriginalName: x
    Type:
      Qualifiers: []
      UnqualifiedType: /Types/TypedefType-10726857336666412012
  ID: 352442133061364510
  Kind: CABIFunctionType
  OriginalName: fn
  ReturnType:
    Qualifiers: []
    UnqualifiedType: /Types/TypedefType-2692631726386566055
```
I will add a test for this.

Reply:
C producer:
```
typedef char uint32_t;

uint32_t fn(uint32_t x)
{
  return x;
}
```